### PR TITLE
Support syntax for private properties and methods

### PR DESCRIPTION
**Pull request by zbuc**
Resolves https://github.com/tree-sitter/tree-sitter-javascript/issues/176

Checklist:
- [ ] All tests pass in CI.
- [x] There are sufficient tests for the new fix/feature.
- [x] Grammar rules have not been renamed unless absolutely necessary.
- [x] The conflicts section hasn't grown too much.
- [ ] The parser size hasn't grown too much (check the value of STATE_COUNT in src/parser.c).

### CODE REVIEW

**Comment by zbuc** [comment]
I used the following test file to validate:

```javascript
class ClassWithPrivateField {
  #privateField
}

class ClassWithPrivateMethod {
  #privateMethod() {
    return 'hello world'
  }
}

class ClassWithPrivateStaticField {
  static #PRIVATE_STATIC_FIELD
}
```

**Comment by aryx** [comment]
Could this be done by just adding an optional('#') in method and field definitions?
What does the official JS grammar say?

**Comment by mjambon** [comment]
Hopefully you can avoid introducing the dynamic conflict by using `optional('#')` like @aryx suggests.

**Comment by zbuc** [comment]
I updated the PR to define `PrivateIdentifier` and allow that for field and method members. I had to change the existing `.public_field_definition` to `.field_definition` because now it can be either public or private. That's what originally led me to define new `private_field_definition` and `private_method_definition` types

**Comment by zbuc** [comment]
Unfortunately this is failing a test case where a private method is called:

```javascript
  class ClassWithPrivateMethod {
    publicMethod() {
      return this.#privateMethod()
    }
  
    #privateMethod() {
      return 'hello world'
    }
  }
```

**Comment by zbuc** [comment]
OK, every test case from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields is passing now.

**Review by maxbrunsfeld — CHANGES_REQUESTED** [review]
Your grammar changes look great AFAICT.

It looks like you didn't add any tests yet. Can you add tests in `test/corpus`?

File: grammar.js [hunk 1]
```diff
@@ -723,7 +723,7 @@ module.exports = grammar({
     member_expression: $ => prec('member', seq(
       field('object', choice($.expression, $.primary_expression)),
       choice('.', '?.'),
-      field('property', alias($.identifier, $.property_identifier))
+      field('property', alias(choice($.identifier, $.private_identifier), $.property_identifier))
```
Comment (maxbrunsfeld) [inline grammar.js:726]: Maybe we should actually differentiate between private and public properties in the public API. What do you think?

If we do want that, I think we could:
* rename `private_identifier` to `private_property_identifier`
* in `_property_name`, move `private_property_identifier` *outside* the `alias()`
* similarly, change this line to:

```js
field('property', choice(
  $.private_property_identifier,
  alias($.identifier, $.property_identifier),
))
```

File: test/corpus/expressions.txt [hunk 2]
```diff
@@ -343,6 +343,7 @@ class Foo {
   static one(a) { return a; };
   two(b) { return b; }
   three(c) { return c; }
+  #four(d) { return d; }
```
Comment (maxbrunsfeld) [inline test/corpus/expressions.txt:346]: It doesn't look like there is a test for a private property in a *member expression* (which I see you added support for in the grammar). Is that indeed allowed?

If so, maybe you could update this line to say `{ return this.#d }`, or something like that.

**Review by maxbrunsfeld — COMMENTED** [review]
I left a couple of questions.

File: grammar.js (hunk 1 above)
Comment (zbuc) [reply grammar.js]: I think that makes a lot of sense. Updated.

**Review by maxbrunsfeld — APPROVED** [review]

**Pull request merged**